Implement the `Get` operation on Groves

whereswaldon opened the pull request:
This PR implements and tests using a `Grove` to fetch a node from disk. It contains a single positive test case, but doesn't have exhaustive tests for all possible error conditions.

As part of this PR, i changed the test suite to work with a `File` interface instead of the `*os.File` type. This allows testing using entirely in-memory implementations. Some tests using actual on-disk nodes would also be valuable, but such tests are slower to run. Much of the changeset in this PR is actually just the effects of creating that `File` type on a great many type signatures.

whereswaldon committed:
Add failing test for grove Get operation

whereswaldon committed:
Implement Get operation on Grove with test

aThorp96 commented on grove/grove.go:
```diff
@@ -68,5 +83,33 @@ func NewWithFS(fs FS) (*Grove, error) {
+	file, err := g.Open(filename)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return nil, false, nil
```
To remove a valid return from an error check, would it be preferable to check this in an if-else? Since `errors.Is` returns false if only one of the arguments is nil, we could do something a little more clear like
```go
if errors.Is(err, os.ErrorNotExist) {
    return nil, false, nil
} else if err != nil {
    return nil, false, fmt.Errorf("failed opening node file \"%s\": %w", filename, err)
}
```

aThorp96 replied:
Not necessarily a hard req for approval, I'd just like to take a valid returns out of an error block if possible

whereswaldon replied:
You know... I didn't realize that `errors.Is` worked that way. It's your turn to teach me something.

To validate this, I checked [the documentation](https://godoc.org/errors#Is), but I don't see this info there. However, I was able to confirm it in [this playground](https://play.golang.org/p/AH-78xGdeC3).

@aThorp96 Where did you pick up this tidbit? I will totally restructure to use this form when i get a minute.

aThorp96 replied:
I had a hunch it had defined behavior when err was nil, and since the documentation didn't help I just checked the source code. I also checked the playgroud to make sure it behaved as expected.

aThorp96 commented on grove/grove_test.go:
```diff
@@ -1,40 +1,123 @@
-func (r fakeFS) Open(path string) (*os.File, error) {
-	return r.files[path], nil
+func (r fakeFS) Open(path string) (grove.File, error) {
+	file, exists := r.files[path]
```
I seem to be leaning all kinds of features with these reviews lol

whereswaldon replied:
Yeah, this is really handy. If you use the single-variable form, it panics if the key is undefined within the map. This way ensures that it doesn't panic.

aThorp96 reviewed (approved):
Everything looks good to me. Just that one style note. I personally don't like to have a valid return inside an error block, but its your call on that @whereswaldon

whereswaldon committed:
Simplify error cases

~athorp96 pointed out that errors.Is can handle being passed
nil, so there was no need to nest the different error cases.

whereswaldon commented:
@aThorp96 Thanks for the catch, I've updated the code to reflect that improvement. I'm going to go ahead and merge now.
